### PR TITLE
use kubebuilder go/v4 template

### PR DESCRIPTION
**Pull request by llamerada-jp**
resolve: #213

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
+run:
```
Comment (peng225) [inline .golangci.yml:1]: The change of this file is in the commit a7b467feb8e71ad2978632580e529dec0b7ab4f1. Is this what you expected?
I thougt that using kubebuilder go/v4 template and enabling lint should be in the different commit.

File: Makefile [hunk 2]
```diff
@@ -73,26 +68,38 @@ fmt: ## Run go fmt against code.
...
 
+.PHONY: test
 test: manifests generate tools fmt vet ## Run tests.
 	$(shell go env GOPATH)/bin/staticcheck ./...
 	go install ./...
 	source <($(SETUP_ENVTEST) use -p env $(ENVTEST_K8S_VERSION)); \
 		go test -race -v -count 1 ./... --timeout=60s
 
+GOLANGCI_LINT = $(shell pwd)/bin/golangci-lint
+GOLANGCI_LINT_VERSION ?= v1.54.2
```
Comment (peng225) [inline Makefile:79]: This version setting should be in [versions.mk](https://github.com/topolvm/pvc-autoresizer/blob/main/versions.mk).

**Review by peng225 — COMMENTED** [review]
- When I ran `go mod tidy` on my environment, go.mod and go.sum was changed. Could you check it?
- Because you deleted `config/default/kustomization.yaml`, [the release procedure](https://github.com/topolvm/pvc-autoresizer/blob/4a59fd7a3a6437fddb93ad7bfbd9d365b0e6c39c/RELEASE.md#bump-version) should be changed. Now step 2 to 6 can be deleted.
  - It will be your help to compare it with [the TopoLVM's release procedure](https://github.com/topolvm/topolvm/blob/main/RELEASE.md#bump-version).

**Review by daichimukai — COMMENTED** [review]
I'd like to see separate commits for change set to migrate from kubebuilder go/v3 to go/v4 and those that are not. My understanding is as follows:

What are related:

- `PROJECT` file,
- moving `main.go` to `cmd/main.go` and related changes,
- comment in `Dockerfile`,
- part of changes in `Makefile`.

What are not related:

- golangci-lint related stuff,
- part of changes in `Dockerfile`,
- CONTRIBUTING.md,
- moving `e2e` to `test/e2e`,
- removing `config` and `hack`,
- changes in `runners`, `metrics`.

---

It seems that the change in the project layout might affect in our code not only `main.go`, but also codes for the controller and the webhook. Shouldn't this be addressed?

> change the layout to accommodate the community request to follow the Standard Go Project Layout by moving the api(s) under a new directory called api, controller(s) under a new directory called internal and the main.go under a new directory named cmd

https://book.kubebuilder.io/migration/v3vsv4#common-changes

File: .golangci.yml (hunk 1 above)
Comment (llamerada-jp) [reply .golangci.yml:1]: Commit splits have been revised. There are two commits related to lint.
https://github.com/topolvm/pvc-autoresizer/pull/236/commits/e9077649d9ee8cb4c343ce1b64d0a139ca282075
https://github.com/topolvm/pvc-autoresizer/pull/236/commits/0caf6cf1bcf7d6bd15a4947ff679171c6c1fc324

File: Makefile (hunk 2 above)
Comment (llamerada-jp) [reply Makefile]: The lint installation and version specification was changed in the following commit.
https://github.com/topolvm/pvc-autoresizer/pull/236/commits/e9077649d9ee8cb4c343ce1b64d0a139ca282075

File: Makefile [hunk 3]
```diff
@@ -73,26 +68,38 @@ fmt: ## Run go fmt against code.
...
+golangci-lint:
```
[the remark this answers is not in the document]
Comment (llamerada-jp) [reply Makefile]: The lint installation and version specification was changed in the following commit.
https://github.com/topolvm/pvc-autoresizer/pull/236/commits/e9077649d9ee8cb4c343ce1b64d0a139ca282075

**Comment by llamerada-jp** [comment]
@peng225 
> * When I ran `go mod tidy` on my environment, go.mod and go.sum was changed. Could you check it?

It was an incomplete work, so I updated it.

> * Because you deleted `config/default/kustomization.yaml`, [the release procedure](https://github.com/topolvm/pvc-autoresizer/blob/4a59fd7a3a6437fddb93ad7bfbd9d365b0e6c39c/RELEASE.md#bump-version) should be changed. Now step 2 to 6 can be deleted.
>   
>   * It will be your help to compare it with [the TopoLVM's release procedure](https://github.com/topolvm/topolvm/blob/main/RELEASE.md#bump-version).

The change was added in the following commit.
https://github.com/topolvm/pvc-autoresizer/pull/236/commits/37b5c6539f89652b083a21b2d79520b5adf583ad

**Comment by llamerada-jp** [comment]
> I'd like to see separate commits for change set to migrate from kubebuilder go/v3 to go/v4 and those that are not. ...

I reworked the split of the commits.

> It seems that the change in the project layout might affect in our code not only `main.go`, but also codes for the controller and the webhook. Shouldn't this be addressed?

Since the official layout was defined differently from the Standard Go Project Layout, I wondered whether I should include it or not. However, the internal/ directory was the same for both, so I changed it. Since there was no specification for constants, I moved it to the top directory in the same as TopoLVM.

File: RELEASE.md [hunk 4]
```diff
@@ -39,36 +39,16 @@ The result should look something like:
 Bump version
 ------------
 
-1. Determine a new version number, and define the `VERSION` variable.
+1. Change `TOPOLVM_VERSION` in `versions.mk` to the latest topolvm chart release tag. (e.g. topolvm-chart-vX.Y.Z)
```
Comment (peng225) [inline RELEASE.md:42]: I'm sorry, but I overlooked that we still needed to update `TOPOLVM_VERSION` in my following comment.
https://github.com/topolvm/pvc-autoresizer/pull/236#pullrequestreview-1815025548

The previous procedure was easy to run because we were able to create a PR almost by copy and pasete. So, I would like you to revert these changes and just delete the step 3 in the previous procedure.

File: hack/boilerplate.go.txt
[the remark this answers is not in the document]
Comment (llamerada-jp) [reply hack/boilerplate.go.txt:1]: added commit message like below.

> `hack/boilerplate.go.txt` file is not generated when execute
`kubebuilder init` command with `--license none` option to keep
the head of each file. And since this project does not use custom
resources, generate-api is not generated either. Therefore, the
related codes are removed.

https://github.com/topolvm/pvc-autoresizer/pull/236/commits/30ad517d3bc66ae8b8a3cf53fbe5e77eadeb54a1

File: internal/hooks/persistentvolumeclaim.go [hunk 5]
```diff
@@ -17,6 +18,7 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
 )
 
+//nolint:lll
```
[the remark this answers is not in the document]
Comment (llamerada-jp) [reply internal/hooks/persistentvolumeclaim.go]: thanks. refrected it.

File: test/e2e/suite_test.go [hunk 6]
```diff
@@ -64,7 +64,7 @@ func kubectlWithInput(input []byte, args ...string) ([]byte, []byte, error) {
 
 func TestMtest(t *testing.T) {
 	if os.Getenv("E2ETEST") == "" {
-		t.Skip("Run under e2e/")
+		t.Skip("Run under test/e2e/")
```
[the remark this answers is not in the document]
Comment (llamerada-jp) [reply test/e2e/suite_test.go:67]: added commit message like below.

> In go/v4 template, files related to test except unit test are
generated in `test/` dir, so move e2e to the same dir.

https://github.com/topolvm/pvc-autoresizer/pull/236/commits/f098f12b3668810cbb88096626ee5f158c8d9e50

File: RELEASE.md (hunk 4 above)
Comment (llamerada-jp) [reply RELEASE.md]: refrected it

File: hack/boilerplate.go.txt
[the remark this answers is not in the document]
Comment (daichimukai) [reply hack/boilerplate.go.txt:1]: Thank you.

**Review by peng225 — APPROVED** [review]

File: test/e2e/suite_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (llamerada-jp) [reply test/e2e/suite_test.go:67]: This did not explicitly appear in the documentation. I read it from the [tutorial](https://book.kubebuilder.io/component-config-tutorial/tutorial) and from the [test data](https://github.com/kubernetes-sigs/kubebuilder/tree/master/testdata/project-v4/test) project.

> Note that most of this tutorial is generated from literate Go files that form a runnable project, and live in the book source directory: [docs/book/src/component-config-tutorial/testdata/project](https://github.com/kubernetes-sigs/kubebuilder/tree/master/docs/book/src/component-config-tutorial/testdata/project).

[the remark this answers is not in the document]
Comment (daichimukai) [reply test/e2e/suite_test.go:67]: Hmm, I see. Thank you.

File: Dockerfile [hunk 7]
```diff
@@ -1,19 +1,17 @@
 # Stage1: Build the pvc-autoresizer binary
-FROM golang:1.20-bullseye as builder
+FROM golang:1.20 as builder
```
Comment (daichimukai) [inline Dockerfile:2]: I don't know why we are using "bullseye" image, but why did you change it?

Comment (llamerada-jp) [reply Dockerfile:2]: > why we are using "bullseye"

There are two images in golang:1.20, bullseye base and bookworm base. When updating from golang:1.19, we specified bullseye explicit for the base image to keep compatibility. 

> why did you change it?

Since pvc-autoresizer programs are static-linked and run in the scratch environment, the base image of the build environment is not an important factor. Therefore, the base image specification is removed to follow the template.

Comment (daichimukai) [reply Dockerfile:2]: Okay I understand that this change is a reasonable one, although not seems to be so much related to migration of kubebuilder. Thank you.

**Review by daichimukai — APPROVED** [review]
